### PR TITLE
8259911: byteArrayViewVarHandle should throw ArrayIndexOutOfBoundsException

### PR DESCRIPTION
**Pull request by cl4es**
Change `MethodHandles.byteArrayViewVarHandle` to throw `ArrayIndexOutOfBoundsException` rather than the more generic `IndexArrayOutOfBoundsException`. This feels more natural, and reduces the risk for subtle behavioral mismatch when migrating code from arrays/Unsafe to VHs. 

CSR: [JDK-8259912](https://bugs.openjdk.java.net/browse/JDK-8259912)

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8259911](https://bugs.openjdk.java.net/browse/JDK-8259911): byteArrayViewVarHandle should throw ArrayIndexOutOfBoundsException

### Reviewers
 * [Jorn Vernee](https://openjdk.java.net/census#jvernee) (@JornVernee - Committer) ⚠️ Review applies to 42c9460ceec829f60c6f33dc43edaa81c5782c99
 * [Mandy Chung](https://openjdk.java.net/census#mchung) (@mlchung - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2124/head:pull/2124`
`$ git checkout pull/2124`

### CODE REVIEW

**Comment by mlchung** [comment]
The change to AIOOBE is reasonable.    Have you considered making this just as an implementation change and leave the spec as is?

**Comment by cl4es** [comment]
> The change to AIOOBE is reasonable. Have you considered making this just as an implementation change and leave the spec as is?

What would be the benefits? AFAICT the CSR is still needed since it's a small behavioral change, and updating the spec helps communicate that change.

**Comment by cl4es** [comment]
I agree the change is allowable within the current spec. The behavior change is observable, though, and however unlikely there might be code that relies on exact class of the exception being thrown.

So while perhaps not strictly needed, the spec change makes sense coming from the other way: If I have some array-based code then that will be throwing AIOOBE on OOBs. But the `byteArrayViewVarHandle` code is specified to throw IOOBE, so a developer picking it up would have to catch and rewrap exceptions or accept the compatibility risk. Harmonizing to specify AIOOBE improves the migration story. Case in point is #1855 which require changes to some tests that are expecting AIOOBE.

**Comment by mlchung** [comment]
> I agree the change is allowable within the current spec. The behavior change is observable, though, and however unlikely there might be code that relies on exact class of the exception being thrown.
> So while perhaps not strictly needed, the spec change makes sense coming from the other way: If I have some array-based code then that will be throwing AIOOBE on OOBs. But the `byteArrayViewVarHandle` code is specified to throw IOOBE, so a developer picking it up would have to catch and rewrap exceptions or accept the compatibility risk. Harmonizing to specify AIOOBE improves the migration story.

I don't think there is any migration story.   The exception type is an improved clarity.   The spec change is very minor and I'm okay with or without the spec change. 

> Case in point is #1855 which require changes to some tests that are expecting AIOOBE.

Yes, that's what I was about to ask too.

File: src/java.base/share/classes/java/lang/invoke/MethodHandles.java [hunk 1]
```diff
@@ -4371,7 +4371,7 @@ public static VarHandle arrayElementVarHandle(Class<?> arrayClass) throws Illega
      * {@code double}.
      * <p>
      * Access of bytes at a given index will result in an
-     * {@code IndexOutOfBoundsException} if the index is less than {@code 0}
+     * {@code ArrayIndexOutOfBoundsException} if the index is less than {@code 0}
```
Comment (JornVernee) [inline src/java.base/share/classes/java/lang/invoke/MethodHandles.java:4374]: Should the copyright year of this file also be updated?

**Review by JornVernee — APPROVED** [review]

**Review by mlchung — APPROVED** [review]
Looks good.  Thanks for updating the test.

**Pull request closed without merging**